ignite/metrics/rouge.py

```diff
@@ -0,0 +1,438 @@
+            if i == 0 or j == 0:
+                dp[i][j] = 0
+            elif seq_a[i - 1] == seq_b[j - 1]:
+                dp[i][j] = dp[i - 1][j - 1] + 1
+            else:
+                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
+
+    return dp[m][n]
+
+
+class Score(namedtuple("Score", ["match", "candidate", "reference"])):
+    def precision(self) -> float:
+        return self.match / self.candidate if self.candidate > 0 else 0
+
+    def recall(self) -> float:
+        return self.match / self.reference if self.reference > 0 else 0
+
+
+def compute_ngram_scores(candidate: Sequence[Any], reference: Sequence[Any], n: int = 4) -> Score:
+    """
+    Compute the score based on ngram co-occurence of sequences of items
+
+    Args:
+        candidate: candidate sequence of items
+        reference: reference sequence of items
+        n: ngram order
+
+    Returns:
+        The score containing the number of ngram co-occurences
+
+    .. versionadded:: 0.5.0
+    """
+
+    # ngrams of the candidate
+    candidate_counter = ngrams(candidate, n)
+    # ngrams of the references
+    reference_counter = ngrams(reference, n)
+    # ngram co-occurences in the candidate and the references
+    match_counters = candidate_counter & reference_counter
+
+    # the score is defined using Fraction
+    return Score(
+        match=sum(match_counters.values()),
+        candidate=sum(candidate_counter.values()),
+        reference=sum(reference_counter.values()),
+    )
+
+
+def compute_lcs_scores(candidate: Sequence[Any], reference: Sequence[Any]) -> Score:
+    """
+    Compute the score based on longest common subsequence of sequences of items
+
+    Args:
+        candidate: candidate sequence of items
+        reference: reference sequence of items
+
+    Returns:
+        The score containing the length of longest common subsequence
+
+    .. versionadded:: 0.5.0
+    """
+
+    # lcs of candidate and reference
+    match = lcs(candidate, reference)
+
+    # the score is defined using Fraction
+    return Score(match=match, candidate=len(candidate), reference=len(reference))
+
+
+class MultiRefReducer(metaclass=ABCMeta):
+    r"""
+    Reducer interface for multi-reference
+    """
+
+    @abstractmethod
+    def __call__(self, scores: Sequence[Score]) -> Score:
+        pass
+
+
+class MultiRefAverageReducer(MultiRefReducer):
+    r"""
+    Reducer for averaging the scores
+    """
+
+    def __call__(self, scores: Sequence[Score]) -> Score:
+        match = sum([score.match for score in scores])
+        candidate = sum([score.candidate for score in scores])
+        reference = sum([score.reference for score in scores])
+        return Score(match=match, candidate=candidate, reference=reference)
+
+
+class MultiRefBestReducer(MultiRefReducer):
+    r"""
+    Reducer for selecting the best score
+    """
+
+    def __call__(self, scores: Sequence[Score]) -> Score:
+        return max(scores, key=lambda x: x.recall())
```

Comment:
Why are we using recall for calculating the best?

Reply:
Also, need to add perl as a dependency to run the perl package I think.

Reply:
> Why are we using recall for calculating the best?

That is the algorithm in the paper.

Reply:
> Also, need to add perl as a dependency to run the perl package I think.
> Check docs here:
> https://github.com/li-plus/rouge-metric
> 
> Also, if you wish to use the perl implementation to compare against I advise against using this module because the perl module is already downloaded and this package will only compare against that fixed version and not any future updates that might happen in the perl module.

Python 3.8.1 failed, others versions seems ok. Last thing to do before merge :)

Reply:
Can you explain why it failed? I thought it failed because of that

Reply:
Also, Looks great! Much cleaner than mine.

Reply:
Ok It is windows and MacOS tests.

Many options are possible for us 
* move to a pythonic ROUGE
* don't execute failing tests on such platforms
* install perl on windows and macos

There is a python version of rouge in `metric-rouge`. I tried it but results seemed different than perl rouge... Maybe we could try this further. 

@gucifer what do you think ?

Reply:
Sure! we can just use pythonic rouge. Let me find one which is most inline with Rouge155